internal/vault/client_factory.go

```diff
@@ -572,12 +661,76 @@ func (m *cachingClientFactory) restoreAllRequest(ctx context.Context, client ctr
 	return req, nil
 }
 
-func (c *cachingClientFactory) incrementRequestCounter(operation string, err error) {
+func (m *cachingClientFactory) incrementRequestCounter(operation string, err error) {
 	if err != nil {
-		c.requestErrorCounterVec.WithLabelValues(operation).Inc()
+		m.requestErrorCounterVec.WithLabelValues(operation).Inc()
 	} else {
-		c.requestCounterVec.WithLabelValues(operation).Inc()
+		m.requestCounterVec.WithLabelValues(operation).Inc()
+	}
+}
+
+func (m *cachingClientFactory) startClientCallbackHandler(ctx context.Context) {
+	logger := m.logger.WithName("clientCallbackHandler")
+	if m.callbackHandlerCancel != nil {
+		logger.Info("Already started")
+		return
 	}
+
+	callbackCtx, cancel := context.WithCancel(ctx)
+	m.callbackHandlerCancel = cancel
+
+	logger.Info("Starting client callback handler")
+
+	go func() {
+		if m.callbackHandlerCh == nil {
+			m.callbackHandlerCh = make(chan Client)
+		}
+		defer func() {
+			m.callbackHandlerCh = nil
+			close(m.callbackHandlerCh)
+		}()
+
+		for {
+			select {
+			case <-callbackCtx.Done():
+				logger.Info("Client callback handler done")
+				return
+			case c := <-m.callbackHandlerCh:
+				if c.IsClone() {
+					continue
+				}
+
+				cacheKey, err := c.GetCacheKey()
+				if err != nil {
+					logger.Error(err, "Invalid client, client callbacks not executed",
+						"cacheKey", cacheKey)
+					continue
+				}
+
+				// remove the client from the cache, it will be recreated when a reconciler
+				// requests it.
+				logger.V(consts.LogLevelDebug).Info("Removing client from cache", "cacheKey", cacheKey)
+				m.cache.Remove(cacheKey)
+				if m.storageEnabled() {
+					if _, err := m.pruneStorage(ctx, m.ctrlClient, cacheKey); err != nil {
+						logger.Info("Warning: failed to prune storage", "cacheKey", cacheKey)
+					}
+				}
+
+				// call in a go routine to avoid blocking the channel
+				go func(c Client) {
+					for idx, cbReq := range m.clientCallbacks {
+						if cbReq.On != ClientCallbackOnLifetimeWatcherDone {
+							continue
+						}
+						logger.Info("Calling client callback on lifetime watcher done",
+							"index", idx, "cacheKey", cacheKey, "clientID", c.ID())
+						cbReq.Callback(ctx, c, nil)
```

Comment:
Remind me when the error parameter of Callback() is used?

Reply:
Seems to be vestigial at this point, I dropped it in 9bb7d311f10eeda0576601e543bf3acaa8c7b2b0